### PR TITLE
Supports checking only initial fragment in url path.

### LINKED ISSUE #49310: Nav Unification: menu highlight disappears in tabbed pages
### Steps to reproduce the behavior

1. Click on a menu item with tabs in the page (ie posts, pages, settings, etc)
2. Click on one of the secondary tabs (ie drafts, scheduled, trashed)
3. See that highlight disappears from sidebar menu

#### What I expected to happen
highlight should stay as it does currently

#### What actually happened
highlight disappears

Note: The problematic areas are noted here paYJgx-1lp-p2

### PR DESCRIPTION
**Pull request by cpapazoglou**
#### Changes proposed in this Pull Request

* Uses a new `itemLinkMatches` function specific for nav unification
* Matches the links by comparing only the first fragment of the path eg `posts`
* Includes a workaround for jetpack types which all start with `types` but differ on the second fragment eg `jetpack-portfolio`
* Includes a temporary fix for sharing buttons being under both tools and settings.

#### Testing instructions

* Navigate around to calypso
* Especially check screens which include secondary navigation eg `pages`, `portfolio`, `tools`
* Make sure that at any case the correct menu is highlighted

**Automated** 
`yarn run test-client client/my-sites/sidebar-unified/test/index.js`

Before | After
-------|------
 | 

Fixes #49310

### CODE REVIEW

File: client/my-sites/sidebar/utils.js [hunk 1]
```diff
@@ -5,18 +5,22 @@ import { includes, some } from 'lodash';
 
 // Check if `path` starts with `pathPrefix` and the very next character
 // (if there is any) is `/` or `?`
-function pathStartsWith( fullPath, pathPrefix ) {
+// If shallow is true, check only the very first fragment.
+function pathStartsWith( fullPath, pathPrefix, shallow ) {
```
Comment (getdave) [inline client/my-sites/sidebar/utils.js]: I'd be tempted to make `shallow` part of an options object. Passing a bool as an arg to a function makes it difficult to understand what's happening without finding and reading the func signature.

File: client/my-sites/sidebar/utils.js [hunk 2]
```diff
@@ -5,18 +5,22 @@ import { includes, some } from 'lodash';
...
+	}
 	return (
 		fullPath === pathPrefix ||
 		( fullPath.slice( 0, pathPrefix.length ) === pathPrefix &&
 			includes( '/?', fullPath[ pathPrefix.length ] ) )
 	);
 }
 
-export function itemLinkMatches( paths, currentPath ) {
+export function itemLinkMatches( paths, currentPath, shallow = false ) {
```
Comment (tjcafferkey) [inline client/my-sites/sidebar/utils.js]: Would be great to update the unit tests to support this.

File: client/my-sites/sidebar/utils.js (hunk 1 above)
Comment (cpapazoglou) [reply client/my-sites/sidebar/utils.js]: I have opted in creating separate function for our unified sidebar.

File: client/my-sites/sidebar/utils.js (hunk 2 above)
Comment (cpapazoglou) [reply client/my-sites/sidebar/utils.js]: Tests incoming!

File: client/my-sites/sidebar-unified/test/index.jsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		} );
+
+		test( 'clicking a marketing panel should activate the marketing/tools menu', () => {
+			const isSelected = itemLinkMatches(
+				'/marketing/tools/example.wordpress.com',
+				'/marketing/traffic/cpapfree.wordpress.com'
+			);
+
+			expect( isSelected ).to.be.true;
+		} );
```
Comment (getdave) [inline client/my-sites/sidebar-unified/test/index.jsx:73]: If these are special edge cases then I'd nest them with their own `describe` block to indicate that they are.

File: client/my-sites/sidebar-unified/utils.jsx [hunk 4]
```diff
@@ -0,0 +1,18 @@
+const pathIncludes = ( currentPath, term, position ) =>
+	currentPath.split( /[/,?]/ )?.[ position ].includes( term );
+
+const fragmentIsEqual = ( path, currentPath, position ) =>
+	currentPath.split( /[/,?]/ )?.[ position ] === path.split( /[/,?]/ )?.[ position ];
+
+// Checks if `currentPath` starts with the first fragment of `path`
```
Comment (getdave) [inline client/my-sites/sidebar-unified/utils.jsx]: Should use proper docblock style here not inline `//`.

File: client/my-sites/sidebar-unified/utils.jsx [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+export const itemLinkMatches = ( path, currentPath ) => {
+	// Accounts for jetpack custom post types, eg portofolio, testimonials.
+	if ( pathIncludes( currentPath, 'types', 1 ) ) {
+		return fragmentIsEqual( path, currentPath, 2 );
+	}
+	// Temp fix till we remove sharing buttons from 'Settings' menu.
```
Comment (getdave) [inline client/my-sites/sidebar-unified/utils.jsx]: Can you add more context here. Why is the "fix" needed? What is it actually fixing?

**Review by getdave — CHANGES_REQUESTED** [review]
Nice work on this. Extra kudos for the tests!

I tested with my site and in general it worked well.

**Update**: I was accidentally testing without having pulled your latest changes. Once I'd pulled down it worked really well. 🚢

**Review by getdave — APPROVED** [review]

File: client/my-sites/sidebar-unified/test/index.jsx (hunk 3 above)
Comment (cpapazoglou) [reply client/my-sites/sidebar-unified/test/index.jsx:73]: done!

File: client/my-sites/sidebar-unified/utils.jsx (hunk 4 above)
Comment (cpapazoglou) [reply client/my-sites/sidebar-unified/utils.jsx]: done!

File: client/my-sites/sidebar-unified/utils.jsx (hunk 5 above)
Comment (cpapazoglou) [reply client/my-sites/sidebar-unified/utils.jsx]: done!
